src/charm.py

```diff
@@ -204,50 +216,81 @@ def _on_clean_up_packages_action(self, event):
         start = time.time()
         logger.info("Cleaning up unreferenced packages")
         packages_to_be_removed, freed_up_space = self._check_packages()
-        prefix_message = "Clean up completed without errors."
-        for package in packages_to_be_removed:
-            try:
-                package.unlink()
-                logger.info("Removed {}".format(package))
-            except Exception as e:
-                logger.error(e)
-                prefix_message = (
-                    "Clean up completed with errors, "
-                    "please refer to juju's log for details."
-                )
+        cleanup_result = clean_packages(packages_to_be_removed)
         elapsed = time.time() - start
-        logger.info("Clean up complete, took {}s".format(elapsed))
 
-        event.set_results(
-            {
-                "message": "{} Freed up {}".format(prefix_message, freed_up_space),
-            }
-        )
+        if cleanup_result is True:
+            message = "Clean up completed without errors."
+        else:
+            message = (
+                "Clean up completed with errors, please refer to juju's log for "
+                "details."
+            )
+
+        logger.info("Clean up complete, took %ds", elapsed)
+        event.set_results({"message": f"{message} Freed up {freed_up_space}"})
+
+    def _get_mirrors(self, source: Optional[str] = None) -> List[str]:
+        """Get filtered mirrors."""
+        mirrors = list(self._stored.config.get("mirror-list", []))
+        logger.debug("found %d mirrors in charm configuration", len(mirrors))
+        if source:
+            reg_filter = re.compile(source)
+            mirrors = list(filter(reg_filter.search, mirrors))
+
+        return mirrors
+
+    def _create_tmp_apt_mirror_config(self, *mirrors) -> Path:
+        """Create tmp apt_mirror config."""
+        with NamedTemporaryFile(delete=False) as tmp_config:
+            tmp_path = Path(tmp_config.name)
+
+        config = dict(self._stored.config).copy()
+        config["mirror-list"] = mirrors
+        self._render_config(config, tmp_path)
+        return tmp_path
 
     def _on_synchronize_action(self, event):
```

Comment:
I'm not sure, that's why I left comment above. I think that maybe both `clean_dists` and `clean_packages` can take filtered list of mirrors into account.
What do you think?

Reply:
I think they should. However, I guess it will be difficult to implement correctly since all you have is  `**/dists` directories and you don't know which part of the directories are coming from which mirror source in the mirror list. That's why the `synchronize_action` was initially designed that way because it's easy to implement it correctly.

Another option is to create a new action that only check if the `source` matches the item in `mirror_list`, and run `apt-mirror <tmp_config_path>` to pull new information; the old information will be replaced.

Reply:
Yes, that's why I did not do it yet. Plus my idea was also to ignore any cleaning if source is provided.

Reply:
Do you have any strong opinion here? Because I do not.
@esunar what do you think?

Reply:
I will vote having a new action. This charm is already an action-based charm (which is a strange charm), adding one more action does not make a difference, and it will make sure old behavior does not change.

Reply:
you kind of do know which directory belongs to which mirror, but resolution may be fragile (and very dangerous). I would be ok with skipping cleaning if synchronizing a subset of all mirrors. TBH this charm should be superseded by Landscape.
I do not however understand @chanchiwai-ray's new action suggestion.

Reply:
As Robert has found, we can correctly clean packages by not removing dists after a filtered synchronization

Reply:
> you kind of do know which directory belongs to which mirror, but resolution may be fragile (and very dangerous)

Yes, that's true. But just like you mention, it's dangerous.

> I do not however understand @chanchiwai-ray's new action suggestion.

The reason for that is, I believe it will take to much effort (I guess) to properly perform a careful clean up with this extra feature. And we kind of cannot get rid of cleaning in the `synchronize` action because of [this bug](https://bugs.launchpad.net/charm-apt-mirror/+bug/1996654); but strictly speak we can remove "cleaning" in `synchronize` action and resort to the `clean-up-packages` action, it will only cause inconvenience for the operator.

However, it seems that @rgildein had a nice and simple solution for that so I will marked this resolved.